### ISSUE TITLE
[Filesystem] LockHandler not injectable because of constructor argument

### ISSUE THREAD

**Issue by parhamdoustdar** [symfony/symfony#15321]
I have a method, `run()` on a class called `CommandRunner`. `CommandRunner::run()` needs to create a lock based on an argument it receives. This makes me unable to inject the `LockHandler` class because the argument of `run()` is actually the constructor argument to `LockHandler`.

In my humble opinion, having a `LockHandlerFactory::create($filename)` will fix this. Of course, I could create this only in my own project, but I thought including this in the component would improve the testability of client code.

**Comment by adamziel** [comment]
Why do you need to inject `LockHandler` instance?

**Comment by adamziel** [comment]
What `LockHandlerFactory::create($filename)` solves that `new LockHandler($filename)` doesn't?

**Comment by parhamdoustdar** [comment]
1. It makes my code more testable: I want to mock my `LockHandler` to return true once, and false the next time. I want to see if `release()` was called or not. i can't do this when my code is `new LockHandler($filename)`.
2. It makes my dependencies clearer: when you look at my constructor, you see that `LockHandlerFactory` is included, which means it has something to do with locking a filesystem. Otherwise, this fact would be hidden from you unless you read the code.

In other words, all the benefits that come with dependency injection would come with having a factory. The way `Lockhandler` currently works is like how the `date()` function in PHP works; you need to create a factory to be able to mock it, or properly test your code.

**Comment by adamziel** [comment]
If I understand correctly, you need to pass an argument to the LockHandler, and the value of this argument is only available in the runtime. So, unless I am missing something important, it is not possible to define it as a service beforehand.

**Comment by adamziel** [comment]
Ah I see, you want to pass the factory instance to the constructor. It makes sense, I could issue a PR with this change if someone closer to the core team approves the idea.

**Comment by nicolas-grekas** [comment]
Closing as the LockHandler class has been deprecated in 3.4 and removed in 4.0.
